lib/improver/nbhood/nbhood.py

```diff
@@ -281,84 +514,159 @@ def _sum_and_area_for_edge_cases(cube, cells_x, cells_y, i, j):
-        return cubelist.merge_cube(), nan_mask
+        # If there is no mask, make a mask of ones. This will speed up the
+        # calculation.
+        if isinstance(cube.data, np.ma.MaskedArray):
+            mask_cube = cube.copy()
+            mask_cube.rename('mask_data')
+            mask_cube.data = np.logical_not(cube.data.mask.astype(int))
+            cube.data = cube.data.data * mask_cube.data
+            cubes_to_sum = iris.cube.CubeList([cube, mask_cube])
+        else:
+            cubes_to_sum = iris.cube.CubeList([cube])
+        return cubes_to_sum
 
-    @staticmethod
-    def run(cube, radius):
+    def _pad_and_calculate_neighbourhood(
+            self, cubes_to_sum, grid_cells_x, grid_cells_y):
+        """
+        Apply neighbourhood processing consisting of the following steps:
+        1. Pad a halo around the input cube to allow vectorised
+           neighbourhooding at edgepoints.
+        2. Cumulate the array along the x and y axes.
+        3. Apply neighbourhood processing to the cumulated array.
+
+        Parameters
+        ----------
+        cubes_to_sum : Iris.cube.CubeList
+            CubeList containing either the input cube, or the input cube and
+            a mask cube.
+        grid_cells_x : Float
+            The number of grid cells along the x axis used to create a square
+            neighbourhood.
+        grid_cells_y : Float
+            The number of grid cells along the y axis used to create a square
+            neighbourhood.
+
+        Returns
+        -------
+        neighhourhood_averaged_cubes : Iris.cube.CubeList
+            CubeList containing the smoothed field after the square
+            neighbourhood method has been applied to either the input cube, or
+            both the input cube and a mask cube.
+        """
+        neighbourhood_averaged_cubes = iris.cube.CubeList([])
+        for cube_to_process in cubes_to_sum:
+            # Pad the iris cube with zeros. This way, the edge effects produced
+            # by the vectorisation of the 4-point method will appear outside
+            # our domain of interest. These unwanted points can be trimmed off
+            # later.
+            cube_to_process = self.pad_cube_with_halo(
+                cube_to_process, grid_cells_x, grid_cells_y)
+            summed_up_cube, nan_masks = self.cumulate_array(
+                cube_to_process)
+            neighbourhood_averaged_cubes.append(
+                self.mean_over_neighbourhood(
+                    summed_up_cube, grid_cells_x, grid_cells_y, nan_masks))
+        return neighbourhood_averaged_cubes
+
+    def _remove_padding_and_mask(
+            self, neighbourhood_averaged_cubes, pre_neighbourhood_cubes,
+            cube_name, grid_cells_x, grid_cells_y):
+        """
+        Remove the halo from the padded array and apply the mask, if required.
+
+        Parameters
+        ----------
+        neighhourhood_averaged_cubes : Iris.cube.CubeList
```

Comment:
TYPO: neighbourhood (might want to search for any more "neighhourhood" instances).